### ISSUE TITLE
Vote resolver allows for multiple votes if you submit too many requests in a short time

### ISSUE THREAD

**Issue by Discordius** [VulcanJS/Vulcan-Starter#31]
The current vote resolver checks the database whether you've already voted on something, and if you have, cancels your vote, which changes the score of the item you are voting on. If you vote again before the database operation completes, then the resolver is going to cancel your vote twice, which means that the score decreases by twice your voting power. This allows you to arbitrarily manipulate the scores of various items. This seems quite bad and we probably want to fix that.

**Comment by SachaG** [comment]
Can you specify how to recreate the bug? Is it just by clicking the upvote button rapidly?

**Comment by Discordius** [comment]
We fixed a bunch of bugs related to this in our codebase. You can take a
look at our Vulcan fork to figure out the differences in implementation:
https://github.com/LessWrong2/Vulcan

**Comment by jdowning100** [comment]
Likely related to this: https://vulcanjs.slack.com/archives/C02L990AF/p1538612081000100

> In particular, vote.js has a call to:
`await Connectors.delete(Votes, {documentId: document._id, userId: user._id});`
But that breaks now, because that gets translated to
`{ userId: <userId>, _id: <documentId> }`
Which is importantly different, and basically breaks all the clearVote functionality

**Comment by jwrubel** [comment]
@Discordius I was able to extract what I needed from your repo and I got voting working correctly - thank you! It was quite a lot of changes and they're in the voting package, not Vulcan-Starter. Still I have what I changed in a single commit so I'm happy to add it as a PR, even if it's just for someone to take a look at and re-implement properly.

**Comment by jdowning100** [comment]
@jwrubel Could you send a link to your commit (if it’s public)?

**Comment by jwrubel** [comment]
Good idea - there's no reason not to make it public so I just opened it. I had originally started trying to keep individual commits but that got too hard, so these four commits should be enough to correct voting functionality: [1](https://github.com/kennenai/Vulcan/commit/1535c3d9d1fab16e399e1f94563a2bac48a82769)  [2](https://github.com/kennenai/Vulcan/commit/ff6ecf5a134cf2cbab8588e6ec299ed229dc8dcf)  [3](https://github.com/kennenai/Vulcan/commit/8889fea302b82b9a7f3d5ee5bd40a696ce0b81b4)  [4](https://github.com/kennenai/Vulcan/commit/b87651e7ef48891e408c68a312d18a87aff3fbf6) . That last one is the big one. To be honest I haven't really traced through all of these changes to understand how they work; It is definitely possible some of these are not needed. But I can confirm these worked  in my implementation. I'm happy to try and assist if you run in to issues implementing these changes, to the extent I'm able.

**Comment by eric-burel** [comment]
Quote from Halima Olapade on Slack that might explain this issue:
"I think i figured out why the inbuilt voting is breaking. It's cause the Vote collection schema has a field called "documentId" but "documentId" seems to be some kind of reserved identifier. As a result this query always fails: https://github.com/VulcanJS/Vulcan/blob/devel/packages/vulcan-voting/lib/modules/vote.js#L52 so the system can't tell that there's a previous upvote that should be annulled""
